pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -0,0 +1,447 @@
+// code and the go routine responsible for managing the node's labels
+type nodeInfoManager struct {
+	nodeName      types.NodeName
+	k8s           kubernetes.Interface
+	csiKubeClient csiclientset.Interface
+}
+
+// If no updates is needed, the function must return the same Node object as the input.
+type nodeUpdateFunc func(*v1.Node) (newNode *v1.Node, updated bool, err error)
+
+// Interface implements an interface for managing labels of a node
+type Interface interface {
+	// Record in the cluster the given node information from the CSI driver with the given name.
+	AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error
+
+	// Remove in the cluster node information from the CSI driver with the given name.
+	RemoveNodeInfo(driverName string) error
+}
+
+// NewNodeInfoManager initializes nodeInfoManager and returns available interfaces
+func NewNodeInfoManager(
+	nodeName types.NodeName,
+	kubeClient kubernetes.Interface,
+	csiKubeClient csiclientset.Interface) Interface {
+	return &nodeInfoManager{
+		nodeName:      nodeName,
+		k8s:           kubeClient,
+		csiKubeClient: csiKubeClient,
+	}
+}
+
+func (nim *nodeInfoManager) AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error {
+	if driverNodeID == "" {
+		return fmt.Errorf("error adding CSI driver node info: driverNodeID must not be empty")
+	}
+
+	if topology == nil {
+		return fmt.Errorf("error adding CSI driver node info: topology must not be nil")
+	}
+
+	err := nim.updateNode(
+		updateNodeIDInNode(driverName, driverNodeID),
+		updateTopologyLabels(topology),
+	)
+	if err != nil {
+		return fmt.Errorf("error updating Node object with CSI driver node info: %v", err)
+	}
+
+	err = nim.updateCSINodeInfo(driverName, driverNodeID, topology)
+	if err != nil {
+		return fmt.Errorf("error updating CSINodeInfo object with CSI driver node info: %v", err)
+	}
+	return nil
+}
+
+func (nim *nodeInfoManager) RemoveNodeInfo(driverName string) error {
+	err := nim.removeCSINodeInfo(driverName)
+	if err != nil {
+		return fmt.Errorf("error removing CSI driver node info from CSINodeInfo object %v", err)
+	}
+
+	err = nim.updateNode(
+		removeNodeIDFromNode(driverName),
+	)
+	if err != nil {
+		return fmt.Errorf("error removing CSI driver node info from Node object %v", err)
+	}
+	return nil
+}
+
+// updateNode repeatedly attempts to update the corresponding node object
+// which is modified by applying the given update functions in order.
+func (nim *nodeInfoManager) updateNode(updateFuncs ...nodeUpdateFunc) error {
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Retrieve the latest version of Node before attempting update, so that
+		// existing changes are not overwritten. RetryOnConflict uses
+		// exponential backoff to avoid exhausting the apiserver.
+
+		nodeClient := nim.k8s.CoreV1().Nodes()
+		node, err := nodeClient.Get(string(nim.nodeName), metav1.GetOptions{})
```

Comment:
are there plans to use a shared informer?  If so, we should clone the node object first before modifying it so it doesn't impact other uses that may be sharing the same pointer.

Reply:
No I wasn't planning to. Driver registration is fairly infrequent so watching the object doesn't seem worth it